### PR TITLE
Feature/qwen eligen support

### PR DESCRIPTION
**Pull request by nolan4**
Pull Request: Add Entity-Level Image Generation (EliGen) for Qwen Image

Summary

This update implements Entity-Level Image Generation (EliGen) for the Qwen Image model, allowing region-specific prompts through spatial masks. The feature provides fine-grained control over image generation by applying separate attention masks for each entity.

Key Features
	•	Spatial attention masking with isolated entity prompts
	•	Automatic mask resizing to match latent dimensions
	•	RoPE embedding implementation aligned with DiffSynth Studio
	•	Support for batch_size > 1
	•	Backward compatible with standard Qwen Image workflows

### CODE REVIEW

File: comfy/ldm/qwen_image/model.py [hunk 1]
```diff
@@ -2,15 +2,139 @@
...
 from comfy.ldm.lightricks.model import TimestepEmbedding, Timesteps
 from comfy.ldm.modules.attention import optimized_attention_masked
 from comfy.ldm.flux.layers import EmbedND
 import comfy.ldm.common_dit
 import comfy.patcher_extension
 
+logger = logging.getLogger(__name__)
+
+
+class QwenEmbedRope(nn.Module):
```
Comment (Kosinkadink) [inline comfy/ldm/qwen_image/model.py]: Review from comfy: he's still like to see this replaced with the rope code that is already in ComfyUI. Let me know if there are issues that would prevent this from being possible!

File: comfy/ldm/qwen_image/model.py [hunk 2]
```diff
@@ -149,18 +153,58 @@ def forward(
...
 
-        joint_query = torch.cat([txt_query, img_query], dim=1)
-        joint_key = torch.cat([txt_key, img_key], dim=1)
-        joint_value = torch.cat([txt_value, img_value], dim=1)
-
-        joint_query = apply_rotary_emb(joint_query, image_rotary_emb)
-        joint_key = apply_rotary_emb(joint_key, image_rotary_emb)
-
+        # Handle both tuple (EliGen) and single tensor (standard) RoPE formats
+        if isinstance(image_rotary_emb, tuple):
```
Comment (comfyanonymous) [inline comfy/ldm/qwen_image/model.py]: This branch is not needed it is equivalent to the original code if you concat the txt and img ropes together.

**Comment by krigeta** [comment]
Hey @nolan4 thank you so much for this amazing work! I am trying to test this out but not able to run it, I am using this PR and using the [Eligen Lora](https://www.modelscope.cn/models/DiffSynth-Studio/Qwen-Image-EliGen-V2/files) provided by Diffsynth Studio and using this [workflow](https://pastebin.com/vnfUAU4R), please check.

File: comfy/ldm/qwen_image/model.py (hunk 2 above)
Comment (krigeta) [reply comfy/ldm/qwen_image/model.py]: Hey, a lot of people, including me, are very interested in this, and it is very needed. Please show some support, mate. Help the community.

**Comment by krigeta** [comment]
hey @nolan4 may you please reply?

**Comment by nolan4** [comment]
Hi @krigeta — here’s a screenshot of my workflow, which is based on the Qwen text-to-image template. I’m also using the same EliGen LoRA from Diffsynth Studio that you linked. Hope this helps you get it running!
eligen example workflow

**Comment by krigeta** [comment]
Hey @nolan4 thank you so much for this, what if this branch is not merged then is it possible to create a custom node of this?

and yeah i will test this and share the results for sure.

**Comment by krigeta** [comment]
hey @nolan4 it is not working in my case, please check.
Screenshot (99)

**Comment by nolan4** [comment]
Looks like you have multiple loras, try just one for testing purposes. I tried your prompt with a few minor modifications:
dragonball z example

**Comment by Kosinkadink** [comment]
test locally and works, comfy will do a code review to see if anything else needs to be changed!

**Comment by krigeta** [comment]
> Looks like you have multiple loras, try just one for testing purposes. I tried your prompt with a few minor modifications: dragonball z example

Will gonna test it asap and this time I am gonna test the official example as shown in the official Eligen lora implementation and one more thing I want to ask:

In the Diffsynth studio repo, they said it is important to make those colored masks with text overlays. Is that true?

As in their official example, the masks are overlaid to achieve smooth results.

**Comment by makisekurisu-jp** [comment]
How can I get this node?

**Comment by krigeta** [comment]
>  How can I get this node?

you have to install this PR and not the actual comfyUI setup to use this node, as of now it is not the part of the main repo.

**Comment by krigeta** [comment]
Hey @nolan4, I guess this implementation is missing the colour-coded masks that help the lora to differentiate between the regions when they overlap. Please look into it.

**Comment by geroldmeisinger** [comment]
btw something similar was implemented in the Inspire Pack called "regional conditioning by color masks" in case you need inspiration or code
https://github.com/ltdrdata/ComfyUI-extension-tutorials/blob/Main/ComfyUI-Inspire-Pack/workflow/RegionalCFG-RegionalConditioning.png
RegionalCFG-RegionalConditioning
https://github.com/ltdrdata/ComfyUI-Inspire-Pack

**Comment by geroldmeisinger** [comment]
> How can I get this node?

```bash
comfy-cli --workspace ./ComfyUI_eligen install --pr "#10473"
comfy-cli --workspace ./ComfyUI_eligen launch
```

**Comment by geroldmeisinger** [comment]
from https://www.modelscope.cn/models/DiffSynth-Studio/Qwen-Image-EliGen code example

global prompt:
`The promotional poster for Qwen-Image-EliGen Magic Cafe features two magical coffee cups, one flaming and the other spitting out ice cones, against a light blue mist background. The poster reads "Qwen-Image-EliGen Magic Cafe" and "New Arrival."`

entity prompts:
`A red magical coffee cup with flames burning inside.`
`A red magical coffee cup surrounded by ice cones.`
`Text: "New Arrival"`
`Text: "Qwen-Image-EliGen Magic Cafe"`

**Comment by geroldmeisinger** [comment]
> Looks like you have multiple loras, try just one for testing purposes. I tried your prompt with a few minor modifications

I can confirm it works: 8step lora -> eligen lora -> ksampler

> I guess this implementation is missing the colour-coded masks that help the lora to differentiate between the regions when they overlap. Please look into it.

I can confirm it works

Screenshot from 2025-11-02 21-01-08

[eligen_colorspheres.json](https://github.com/user-attachments/files/23292615/eligen_colorspheres.json)

mask_ball1
mask_ball2
mask_ball3


note i use euler+beta, cfg=1, 1328x1328 but that shouldn't make much difference

**Comment by krigeta** [comment]
Hey @geroldmeisinger , thank you so much for the share, is there any other social media where we can chat?

**Comment by geroldmeisinger** [comment]
here are the original masks https://www.modelscope.cn/datasets/DiffSynth-Studio/examples_in_diffsynth/files `data/examples/eligen/entity_control`
of the eligen demo page https://www.modelscope.cn/models/DiffSynth-Studio/Qwen-Image-EliGen

**Comment by krigeta** [comment]
> here are the original masks https://www.modelscope.cn/datasets/DiffSynth-Studio/examples_in_diffsynth/files `data/examples/eligen/entity_control` of the eligen demo page https://www.modelscope.cn/models/DiffSynth-Studio/Qwen-Image-EliGen

These are the ones I got to know about Eligen. In my case, the overlapped masking is not working properly, as I want to create two characters in front of each other, and the view is from the back. Or I would say I use one entity for a character and another for another character, and with ControlNet? Does it work?

And if possible, may you share your Discord or other social app?

File: comfy/model_base.py [hunk 3]
```diff
@@ -119,7 +119,6 @@ def convert_tensor(extra, dtype, device):
             extra = comfy.model_management.cast_to_device(extra, device, None)
     return extra
 
-
```
Comment (Kosinkadink) [inline comfy/model_base.py:122]: Restore the whitespace to prevent extra changes in the PR.

File: comfy/ldm/qwen_image/model.py [hunk 4]
```diff
@@ -149,18 +152,36 @@ def forward(
...
         joint_value = torch.cat([txt_value, img_value], dim=1)
 
+        # Apply RoPE to concatenated queries and keys
         joint_query = apply_rotary_emb(joint_query, image_rotary_emb)
         joint_key = apply_rotary_emb(joint_key, image_rotary_emb)
 
+        # Apply EliGen attention mask if present
+        effective_mask = attention_mask
+        if transformer_options is not None:
+            eligen_mask = transformer_options.get("eligen_attention_mask", None)
```
Comment (Kosinkadink) [inline comfy/ldm/qwen_image/model.py]: Assuming the attention_mask input already works with the way the eligen_attention_mask is passed into the blocks, can you remove this from being redundantly in transformer_options in the whole chain of code?

File: comfy/ldm/qwen_image/model.py [hunk 5]
```diff
@@ -2,15 +2,19 @@
...
-from einops import repeat
+from einops import repeat, rearrange
 
 from comfy.ldm.lightricks.model import TimestepEmbedding, Timesteps
 from comfy.ldm.modules.attention import optimized_attention_masked
 from comfy.ldm.flux.layers import EmbedND
 import comfy.ldm.common_dit
 import comfy.patcher_extension
 
+logger = logging.getLogger(__name__)
```
Comment (Kosinkadink) [inline comfy/ldm/qwen_image/model.py]: Use logging directly instead (```logging.info```, ```logging.debug```, etc.)

**Comment by geroldmeisinger** [comment]
image

**Comment by krigeta** [comment]
> image

Haha indeed, they are needed btw, I want to ask, are you able to get the result shown here?

image

image

**Comment by nolan4** [comment]
@geroldmeisinger @krigeta

Latest version supports 8 entities! Below is a screenshot for the referenced DiffSynth [example](https://github.com/modelscope/DiffSynth-Studio/blob/538017177a6136f45f57cdf0b7c4e0d7e1f8b50d/examples/qwen_image/model_inference_low_vram/Qwen-Image-EliGen.py#L100). 

reference example

**Comment by geroldmeisinger** [comment]
awesome!
see https://docs.comfy.org/custom-nodes/backend/more_on_inputs#dynamically-created-inputs in case you want to make it dynamic

@krigeta https://github.com/krigeta/eligen_test/issues/1

**Comment by Kosinkadink** [comment]
There will be a native growing input type soon - any custom javascript to do this will not be allowed in core outside of a 'general' implementation! That will allow an indefinite amount of inputs into the node.

**Comment by krigeta** [comment]
Hi @nolan4 amazing work, and I hope it will be merged asap.
Hi @Kosinkadink hope the native growing input type will be a part, and really, really thank you both for this, guys!

**Comment by Ling-APE** [comment]
Hi, thanks for implementing this, I'm assuming this implementation will not work with Sageattention or Nunchaku since it is using Diffsynth's pipeline and they will likely to bypass the attention masking logic?

I have successfully reproduced the result using the workflow provided here and also tested Eligen V2:

WhatsApp 圖片2025-11-1820 02 27_6f98a9da
WhatsApp 圖片2025-11-1820 02 52_8049f144

Here is the results from V1 :

WhatsApp 圖片2025-11-1820 04 11_c0317e3a
WhatsApp 圖片2025-11-1820 04 05_3b004a08

Here is the result with Nunchaku:
WhatsApp 圖片2025-11-1820 12 58_c4a0d50b

Here is the result with Sageattention:

WhatsApp 圖片2025-11-1820 17 29_0ab1f199

**Comment by sparedevi** [comment]
hey @nolan4 a lot is going on, but is there any update on a native growing input type?

**Comment by Kosinkadink** [comment]
Autogrow support will be in core stable within a couple weeks: https://github.com/Comfy-Org/ComfyUI_frontend/pull/7275

I'll chaperone this PR into making it into core

**Comment by krigeta** [comment]
Hyey @Kosinkadink I know there is a lot going on but when this PR will be the part of core?

**Comment by krigeta** [comment]
hey @nolan4 is there any update that this will be merged?
